### ISSUE TITLE
Incorrect theory axiom in prime proof

### ISSUE THREAD

**Issue by gebner** [gapt/gapt#499]
Looking through the prime proof, there seems to be at least one incorrect theory axiom (during proof construction, I think it becomes valid again because we substitute with `[k\0]`).  Running `examples/runMain at.logic.gapt.examples.prime.prime3` in the `safe-theory` branch outputs the following:
```
(set-logic QF_UFNIA)
(set-option :produce-unsat-cores true)
(declare-fun f1 () Int)
(declare-fun f2 () Int)
(declare-fun f3 () Int)
(declare-fun f4 () Int)
(declare-fun f5 () Int)
(assert (or (< 0 f1) (= 0 f1)))
(assert (or (< 0 f2) (= 0 f2)))
(assert (or (< 0 f3) (= 0 f3)))
(assert (or (< 0 f4) (= 0 f4)))
(assert (or (< 0 0) (= 0 0)))
(assert (or (< 0 f5) (= 0 f5)))
(assert (! (< f1 f4) :named a0))
(assert (! (< f1 f3) :named a1))
(assert (! (= (+ f1 (* f2 f3)) (+ f4 (* f5 f3))) :named a2))
(assert (! (< 0 f3) :named a3))
(check-sat)
-> sat

Wrong theory axiom: i < k, i < l, i + n * l = k + n_0 * l, 0 < l :- 
```

Z3 outputs the following counterexample (this maps to `i=1`, `k=3`, `l=2`, `n=2`, `n_0=1`):
```
(model 
  (define-fun a2 () Bool
    true)
  (define-fun a1 () Bool
    true)
  (define-fun f3 () Int
    2)
  (define-fun f4 () Int
    3)
  (define-fun f2 () Int
    2)
  (define-fun a0 () Bool
    true)
  (define-fun f1 () Int
    1)
  (define-fun a3 () Bool
    true)
  (define-fun f5 () Int
    1)
)
```

This happens [here](https://github.com/gapt/gapt/blob/274ba894602abff8fb0005577d1dcfefb9db929d/examples/prime/prime.scala#L227) in the `progClosed` lemma.

**Comment by loewenheim** [comment]
Yeah, that one was already weird in the hlk script. Have a look at line 169 in prog_closed.hlk. The end sequent of `prog-closed-p212` contains the formula `k < l` that seems to come out of nowhere.

**Comment by gebner** [comment]
Hmm, the axiom in the HLK file seems to be correct though (rearranged for better visibility):
```
i_0 < k, k < l, i_0 < l, i_0 + q_0 * l = k + r_0 * l :-   (HLK, correct)
 i  < k, 0 < l,  i  < l,  i  +  n  * l = k + n_0 * l :-  (current prime proof)
```
In the current version, the `k<l` has been replaced by `0<l`.  Although I'm not really clear why we actually talk about a generic `k` here--`k` always seems to be substituted by `0` in the end.

**Comment by loewenheim** [comment]
Sure it's correct, but where does the `0 < l` from the previous proof go and where does the `k < l` come from? Am I not understanding something?

**Comment by gebner** [comment]
Oh, now I see what you mean.  Or rather, I don't see any `k < l` in `prog-closed-p21`...

**Comment by loewenheim** [comment]
My point exactly. So I said "screw it" and duplicated the proof as closely as I could ^^

**Comment by gebner** [comment]
Fixed by @loewenheim in b4966792cd3a54e1db969361356c9ae37c8c4a2f
